Tests/SwiftGRPCNIOTests/NIOFunctionalTests.swift

```diff
@@ -0,0 +1,266 @@
+/*
+ * Copyright 2018, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import Dispatch
+import Foundation
+import NIO
+import NIOHTTP1
+import NIOHTTP2
+@testable import SwiftGRPCNIO
+import XCTest
+
+protocol NIOFunctionalTests: class {
```

Comment:
What's the advantage of putting these into a protocol?

Reply:
We can use `Self` in the default implementation:

```swift
static var allTests: [(String, (Self) -> () throws -> Void)]
```

If we didn't then I think we'd have to override this in each subclass.

I was also thinking we should use `swift test --generate-linuxmain` or similar script instead but wanted to save that for a different PR.

Reply:
Ah, thanks for the elaboration. Let's stick with the protocol in this PR and run `swift test --generate-linuxmain` in a follow-up PR (possibly adding it to the `Makefile`; we could even check that this doesn't update anything as part of CI).